### ISSUE TITLE
Add child::async_wait

### ISSUE THREAD

**Issue by RedDwarf69** [boostorg/process#57]
It would be very helpful to have an async_wait() method in `child`. In Unix it can be done asynchronous waiting for the SIGCHLD signal. I'm guessing there is something similar in Windows.

**Comment by klemens-morgenstern** [comment]
I decided to not require `boost.asio for the usage of `booost.process`, so the current design doesn't cut it. You can use `on_exit` for that, which uses `asio` and `SIGCHLD` under the surface. Does that not cut it for you?

**Comment by RedDwarf69** [comment]
I didn't actually noticed on_exit was there, sorry. Yes, it's what I need.

**Comment by RedDwarf69** [comment]
I actually have some questions...

In https://www.boost.org/doc/libs/1_68_0/doc/html/boost/process/group.html#idp44884880-bb "Wait for the process group to exit" means "wait for any process of the process group to exit" or "wait for all the processes of the process group to exit"? Whatever the case, maybe the docs could be modified to make it extra clear?

I would want to be notified when my child, and every descendant of my child, has exited. This is probably only doable in Linux if cgroups are used, so limiting a bit my request (even if it would be cool to use cgroups to support this): I would want to be notified when my child, and every descendant of my child still in the same process group, has exited. This is not really doable with boost.proccess currently, is it?

Does the behaviour of https://www.boost.org/doc/libs/1_68_0/doc/html/boost/process/on_exit.html depend on whether the child was created with a process group or not? There is a need for an "on_group_exit"?

**Comment by klemens-morgenstern** [comment]
1. Yeah, it means all children of the group. Rephrasing might be nice.
2. I am not sure that's generally possible, since deep nested child processes might be in another group or have a `SIGIGN` set. You can just get the descriptor of the group through `native_handle` to use it for extensions - if that cuts it for cgroup.
3. I am not sure that's feasible, because there's no `SIGGRP` - how would you implement that?

**Comment by RedDwarf69** [comment]
Yeah, at least to start with I would not bother trying to handle children that have changed process group. Those get lost. But I don't think it matters whether SIGCHLD is SIG_IGN in the children themselves, you will still get a signal if your own SIGCHLD is not SIG_IGN. If some child in the hierarchy has SIGCHLD with SIG_IGN either:
- The grandchildren will clean up by themselves
- The child will die first, the grandchildren will become your children, and you will waitid() for them.

I guess "on_exit" currently works like this:
1) Wait for SIGCHLD
2) Waitid for the children that SIGCHLD has just signalled

"on_group_exit" or "on_exit when the child has been created with a group" could:
1) Wait for SIGCHLD
2) waitid() for the children that SIGCHLD has just signalled
3) (async)Wait for SIGCHLD
4) waitid() for the process group id using the WNOHANG option
5) If `4` returns ECHILD it means all the children in the process group have exited, `3` is cancel() and "on_group_exit"/"on_exit" is called

**Comment by RedDwarf69** [comment]
Not sure I follow you here. But probably what you refer to can be solved by using WNOWAIT, in addition to WNOHANG, in `4`?

http://man7.org/linux/man-pages/man2/waitpid.2.html
`WNOWAIT     Leave the child in a waitable state; a later wait call can be used to again retrieve the child status information.`

**Comment by klemens-morgenstern** [comment]
Ok, so you have process with pid `12` in group `13` and you wait for both asynchronously. If `SIGCHLD` gets triggered, the waiting for 12 gets it, calls `waitpid(12, &status, WNOHANG)` to check if it got it. Then it does not forward the signal. Even if it does, the handler for group 13 gets called, checks with `waitpid(-13, &status, WNOHANG)` but then gets nothing because the information already got handled by the first call. That goes both ways, so we lose information.

**Comment by RedDwarf69** [comment]
>     1. Yeah, it means all children of the group. Rephrasing might be nice.

FWIW I was now ready to open a different issue about this, since I was testing with boost 1.68. But I see you have fixed it already for 1.69 in https://github.com/boostorg/process/commit/eea73753b572ec5df69b2a624cbacf7569fde835 and it uses a quite similar thing to what I have just mentioned.
